mkchromecast.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" (C) Copyright 2017 Muammar El Khatib <muammar@debian.org>,
.\"
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH Mkchromecast "1" "December 23 2017"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
mkchromecast \- cast audio and video to Google cast devices
.SH SYNOPSIS
.B mkchromecast
.RI [ options ]
.SH DESCRIPTION
This manual page documents briefly the
.B mkchromecast
commands.
.PP
.\" TeX users may be more comfortable with the \fB<whatever>\fP and
.\" \fI<whatever>\fP escape sequences to invode bold face and italics,
.\" respectively.
\fBmkchromecast\fP is a program that allows you to cast your audio or video
files to your Google cast or Sonos devices.
.SH OPTIONS
This program follows the usual GNU command line syntax, with long
options starting with two dashes (`-').
A summary of options is included below.
For a complete description, see the Info files.
.TP
.B \-h, \-\-help
Show summary of options.
.TP
.B \-v, \-\-version
Show version of program.
.SH SEE ALSO
.\".BR bar (1),
.\".BR baz (1).
.\".br
The program is documented fully executing
.IR "mkchromecast -h" ,
that should give you complete access to the usage.


changelog.md
* Mkchromecast (0.3.9) **unreleased**

    - New flag `--loop` to loop video indefinitely while streaming. Closes
      #113.
    - New flag `--mtype` to specify media type. Closes #128.
    - Modified how pyqt is used to accomodate pyqt 5.5, which is what comes in
      Ubuntu 16.04 LTS
    - Removed the -max_muxing_queue_size ffmpeg flag because it's not supported in ffmpeg 2.8.11.
    - Screencast with audio.
    - 10bits mkv files are now encoded to 8bits. This fixes #156.
    - node streaming updated to work with `node v9.8.0`.
    - Improved pause/resume when casting videos. Closes #97.
    - Added support for Fedora packages. Thanks to @leamas.
    - setup.py Linux support. Closes #173. Thanks to @leamas.

* mkchromecast (0.3.8.1) **2017/12/24**

    - A bug when no devices where found has been fixed in this release.

* mkchromecast (0.3.8) **2017/12/23**

    - Improved stability.
    - The macOS bundle is now renamed with capital M, and will not be showing
      in the dock.
    - Chunk size changed from 1024 to 64. Added two more variables that
      decreases the delay considerably.
    - Improved ffmpeg commands for pulseaudio part of the code.
    - node streaming updated to work with `node v9.3.0`.
    - Manpage is now shipped in main branches.
    - TypeError caused by a print statement for Soco devices has been fixed.
      Closes #80.
    - Added Opus codec support.
    - It is now possible to screencast using mkchromecast.
    - Using the `--youtube` flag works with all supported websites by youtube-dl.
    - Correct signal handling using the `signal` module. PR #87 by @Foxboron.
    - Renamed `--reconnect` to `--hijack`. Closes #25.
    - New flag `tries` to to limit the number of tries to connect to
      a chromecast. Closes #54.
    - Allow custom server port with ffmpeg or avconv. Related to #122.
    - Error with "width not divisible by 2 (853x480)". Closes issue #119.
    - The `segment_time` flag has been fixed. Closes issue #71.
    - New flag `command` for setting a custom ffmpeg command. Closes issue
      #109.

* mkchromecast (0.3.7.1) **2017/05/21**

    - macOS bundle built in Yosemite to add more compatibility.
    - Bumped version to match debian official release.

* mkchromecast (0.3.7) **2017/05/20**

    - node streaming updated to work with `node v7.10.0`.
    - Added ALSA device configuration in preferences pane.
    - Improved systray stability, and usability.
    - gstreamer support for capturing audio.
    - Fixed problem related to setting `ogg` and `aac` bitrate. Closes #21.
    - A `--segment-time` option added to modify the segment time when using
      ffmpeg.
    - A `--reconnect` option that monitors if mkchromecast has disconnected
      from google cast, and tries to reconnect.
      Closes [#25](https://github.com/muammar/mkchromecast/issues/25).
    - ALSA device can be set from systray.
    - Add support for newer `pychromecast` versions.
      Closes [#32](https://github.com/muammar/mkchromecast/pull/32).
    - Making ping code python 3 compatible. Closes:
      [#35](https://github.com/muammar/mkchromecast/pull/35).
    - Fixed problem when having various Google cast devices. Closes
      [#50](https://github.com/muammar/mkchromecast/issues/50)
    - Added support to 192000Hz sampling rate support (Closes:
      [#39](https://github.com/muammar/mkchromecast/issues/39)).
    - Fixed a minimal problem with size of preferences pane.
    - Video support.
    - node is a supported backend for streaming video.
    - Added Sonos speakers support.
    - FLAC codec supports bitrate.

* mkchromecast (0.3.6) **released**: 2016/09/19

    - The node streaming has been updated to work with `node v6.6.0`.
    - `Ctrl-C` now raises `KeyboardInterrupt` when using `--volume` option from
      console.
    - Improvements under the hood.
    - Now mkchromecast does not need pulseaudio to cast!.
    - You can play from stream using the `--source-url` flag.
    - `-d` option has been enable to discover available Google cast devices.
    - `--host` flag allows users to manually enter the local ip. Closes
      [#17](https://github.com/muammar/mkchromecast/issues/17).
    - `-n` flag allows users to pass the name of their cast devices.

* mkchromecast (0.3.5) **released**: 2016/08/26

    - Added close button for preferences pane. Closes
      [#13](https://github.com/muammar/mkchromecast/issues/13)
    - Improvements for cases where there are virtual network interfaces
      present.
    - Dropped `-re` flag from ffmpeg and avconv commands.
    - Renamed `ffmpeg.py` to `audio.py`.

* mkchromecast (0.3.4) **released**: 2016/08/19

    - New white icons.
    - Added 96000Hz sampling rate support for `ffmpeg` and `avconv` backends.
      Closes [#11](https://github.com/muammar/mkchromecast/issues/11).
    - Fixed 48000Hz sample rate case.
    - The node streaming has been updated to work with `node v6.4.0`.

* mkchromecast (0.3.3) **released**: 2016/08/16

    - Improved MultiRoom support. Closes
      [#8](https://github.com/muammar/mkchromecast/issues/8).

* mkchromecast (0.3.2) **released**: 2016/06/15

    - Improvements for cases where chromecasts have non-ascii characters in
      their names. Closes
      [#7](https://github.com/muammar/mkchromecast/issues/7).

* mkchromecast (0.3.1) **released**: 2016/08/12

    - Improved Preferences window.
    - The node streaming has been updated to work with `node v6.3.1`.
    - Improvements in pulseaudio.py for preventing subprocess.Popen from
      displaying output.
    - Improvements in `Check For Updates` method.
    - Added new option `--chunk-size` to control chunk's size of flask server
      when streaming using `ffmpeg` or `avconv`.

* mkchromecast (0.3.0) **released**: 2016/07/12

    - Youtube URLs can be played piping the audio using `youtube-dl`.
    - New method for discovering local IP in macOS. This will improve the
      stability of the system tray application. Closes

mkchromecast.py
#!/usr/bin/env python

# This file is part of mkchromecast.

import mkchromecast.__init__
from mkchromecast.version import __version__
from mkchromecast.audio_devices import (inputint, inputdev, outputdev,
                                        outputint)
from mkchromecast.cast import casting
import mkchromecast.colors as colors
from mkchromecast.pulseaudio import create_sink, remove_sink
from mkchromecast.utils import terminate, checkmktmp, writePidFile
import subprocess
import atexit
import signal


class mk(object):
    """Class to manage cast process"""
    def __init__(self):
        print(colors.bold('Mkchromecast ')+'v'+__version__)
        self.platform = mkchromecast.__init__.platform
        self.adevice = mkchromecast.__init__.adevice
        self.debug = mkchromecast.__init__.debug
        self.ccname = mkchromecast.__init__.ccname
        self.videoarg = mkchromecast.__init__.videoarg
        self.youtubeurl = mkchromecast.__init__.youtubeurl
        self.tray = mkchromecast.__init__.tray
        self.discover = mkchromecast.__init__.discover
        self.source_url = mkchromecast.__init__.sourceurl
        self.encoder_backend = mkchromecast.__init__.backend
        self.select_cc = mkchromecast.__init__.select_cc
        self.control = mkchromecast.__init__.control

        self.cc = casting()
        checkmktmp()
        writePidFile()

        """
        Initializing backend array
        """
        self.backends = [
          'ffmpeg',
          'avconv',
          'parec',
          'gstreamer'
          ]

        self.check_connection()
        if self.tray is False and self.videoarg is False:
            if self.platform == 'Linux':
                self.audio_linux()
            else:
                self.audio_macOS()
        elif self.tray is False and self.videoarg is True:
            self.cast_video()
        else:
            self.start_tray()

    def audio_linux(self):
        """This method manages all related to casting audio in Linux"""
        if self.youtubeurl is None and self.source_url is None:
            if self.adevice is None:
                print('Creating Pulseaudio Sink...')
                print(colors.warning('Open Pavucontrol and Select the '
                      'Mkchromecast Sink.'))
                create_sink()
            print(colors.important('Starting Local Streaming Server'))
            print(colors.success('[Done]'))
            self.start_backend(self.encoder_backend)
            self.cc.initialize_cast()
            self.get_cc(self.select_cc)
            self.cc.play_cast()
            self.show_control(self.control)

        elif self.youtubeurl is None and self.source_url is not None:
            self.start_backend(self.encoder_backend)
            self.cc.initialize_cast()
            self.get_cc(self.select_cc)
            self.cc.play_cast()
            self.show_control(self.control)

        # When casting youtube url, we do it through the audio module
        elif self.youtubeurl is not None and self.videoarg is False:
            import mkchromecast.audio
            mkchromecast.audio.main()
            self.cc.initialize_cast()
            self.get_cc(self.select_cc)
            self.cc.play_cast()
            self.show_control(self.control)

    def audio_macOS(self):
        """This method manages all related to casting audio in macOS"""
        if self.youtubeurl is None and self.source_url is None:
            self.start_backend(self.encoder_backend)
            self.cc.initialize_cast()
            self.get_cc(self.select_cc)

            print('Switching to SoundFlower...')
            inputdev()
            outputdev()
            print(colors.success('[Done]'))
            self.cc.play_cast()
            self.show_control(self.control)

        elif self.youtubeurl is None and self.source_url is not None:
            self.start_backend(self.encoder_backend)
            self.cc.initialize_cast()
            self.get_cc(self.select_cc)
            self.cc.play_cast()
            self.show_control(self.control)

            print('Switching to SoundFlower...')
            inputdev()
            outputdev()
            print(colors.success('[Done]'))
            self.cc.play_cast()
            self.show_control(self.control)

        # When casting youtube url, we do it through the audio module
        elif self.youtubeurl is not None and self.videoarg is False:
            import mkchromecast.audio
            mkchromecast.audio.main()
            self.cc.initialize_cast()
            self.get_cc(self.select_cc)
            self.cc.play_cast()
            self.show_control(self.control)

    def cast_video(self):
        """This method launches video casting"""

        if self.platform == 'Linux':
            print('Creating Pulseaudio Sink...')
            print(colors.warning('Open Pavucontrol and Select the '
                  'Mkchromecast Sink.'))
            create_sink()

        print(colors.important('Starting Video Cast Process...'))
        import mkchromecast.video
        mkchromecast.video.main()
        self.cc.initialize_cast()
        self.get_cc(self.select_cc)
        self.cc.play_cast()
        self.show_control(self.control)

    def get_cc(self, select_cc):
        """Get chromecast name, and let user select one from a list if
        select_cc flag is True.
        """
        # This is done for the case that -s is passed
        if select_cc is True:
            self.cc.sel_cc()
            self.cc.inp_cc()
            self.cc.get_cc()
        else:
            self.cc.get_cc()

    def start_backend(self, encoder_backend):
        """Starting backends"""
        if encoder_backend == 'node' and self.source_url is None:
            from mkchromecast.node import stream
            stream()
        elif encoder_backend in self.backends and self.source_url is None:
            import mkchromecast.audio
            mkchromecast.audio.main()

    def check_connection(self):
        """Check if the computer is connected to a network"""
        if self.cc.ip == '127.0.0.1':        # We verify the local IP.
            print(colors.error('Your Computer is not Connected to Any '
                  'Network'))
            terminate()
        elif self.cc.ip != '127.0.0.1' and self.discover is True:
            self.cc.initialize_cast()
            terminate()

    def terminate_app(self):
        """Terminate the app (kill app)"""
        self.cc.stop_cast()
        if self.platform == 'Darwin':
            inputint()
            outputint()
        elif self.platform == 'Linux':
            remove_sink()
        terminate()

    def controls_msg(self):
        """Messages shown when controls is True"""
        print('')
        print(colors.important('Controls:'))
        print(colors.important('========='))
        print('')
        print(colors.options(           'Volume Up:')+' u')
        print(colors.options(         'Volume Down:')+' d')
        if self.videoarg is True:
            print(colors.options(       'Pause Casting:')+' p')
            print(colors.options(      'Resume Casting:')+' r')
        print(colors.options('Quit the Application:')+' q or Ctrl-C')
        print('')

    def show_control(self, control):
        """Method to show controls"""
        if self.control is True:
            from mkchromecast.getch import getch

            self.controls_msg()
            try:
                while(True):
                    key = getch()
                    if(key == 'u'):
                        self.cc.volume_up()
                        if self.encoder_backend == 'ffmpeg':
                            if self.debug is True:
                                self.controls_msg()
                    elif(key == 'd'):
                        self.cc.volume_down()
                        if self.encoder_backend == 'ffmpeg':
                            if self.debug is True:
                                self.controls_msg()
                    elif(key == 'p'):
                        if self.videoarg is True:
                            print('Pausing Casting Process...')
                            action = 'pause'
                            self.backend_handler(action, self.encoder_backend)
                            if self.encoder_backend == 'ffmpeg':
                                if self.debug is True:
                                    self.controls_msg()
                        else:
                            pass
                    elif(key == 'r'):
                        if self.videoarg is True:
                            print('Resuming Casting Process...')
                            action = 'resume'
                            self.backend_handler(action, self.encoder_backend)
                            if self.encoder_backend == 'ffmpeg':
                                if self.debug is True:
                                    self.controls_msg()
                        else:
                            pass
                    elif(key == 'q'):
                        print(colors.error('Quitting application...'))
                        self.terminate_app()
                    elif(key == '\x03'):
                        raise KeyboardInterrupt
                        atexit.register(self.terminate_app())
            except KeyboardInterrupt:
                self.terminate_app()

        else:
            if self.platform == 'Linux' and self.adevice is None:
                print(colors.warning('Remember to open pavucontrol and select '
                      'the mkchromecast sink.'))
            print('')
            print(colors.error('Ctrl-C to kill the Application at any Time'))
            print('')
            signal.signal(signal.SIGINT,
                          lambda *_: atexit.register(self.terminate_app()))
            signal.signal(signal.SIGTERM,
                          lambda *_: atexit.register(self.terminate_app()))
            signal.pause()

    def backend_handler(self, action, backend):
        """Methods to handle pause and resume state of backends"""
        if action == 'pause' and backend == 'ffmpeg':
            subprocess.call(['pkill', '-STOP', '-f', 'ffmpeg'])
        elif action == 'resume' and backend == 'ffmpeg':
            subprocess.call(['pkill', '-CONT', '-f', 'ffmpeg'])
        elif (action == 'pause' and backend == 'node' and
              self.platform == 'Linux'):
            subprocess.call(['pkill', '-STOP', '-f', 'nodejs'])
        elif (action == 'resume' and backend == 'node' and
              self.platform == 'Linux'):
            subprocess.call(['pkill', '-CONT', '-f', 'nodejs'])
        elif (action == 'pause' and backend == 'node' and
              self.platform == 'Darwin'):
            subprocess.call(['pkill', '-STOP', '-f', 'node'])
        elif (action == 'resume' and backend == 'node' and
              self.platform == 'Darwin'):
            subprocess.call(['pkill', '-CONT', '-f', 'node'])

        if action == 'pause':
            self.cc.pause()
        elif action == 'resume':
            self.cc.play()

    def start_tray(self):
        """This method starts the system tray"""
        import mkchromecast.systray
        checkmktmp()
        writePidFile()
        mkchromecast.systray.main()


if __name__ == "__main__":
    mk()
